### ISSUE TITLE
Do you have a plan to implement polyak averaging?

### ISSUE THREAD

**Issue by dhgrs** [chainer/chainer#4556]
Polyak averaging which is proposed in [1] is used in Parallel WaveNet[2] and other techniques like [3]. Now chainer does not have polyak averaging API. (FYI: TensorFlow has it: https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage) Then do you have a plan to implement polyak averaging?

FYI: I implemented a naive working sample without test code. It is a Classifier-like API.
https://github.com/dhgrs/chainer-ExponentialMovingAverage

[1] Boris T Polyak and Anatoli B Juditsky. Acceleration of stochastic approximation by averaging. SIAM Journal on Control and Optimization, 30(4):838–855, 1992.
[2] OORD, Aaron van den, et al. Parallel WaveNet: Fast High-Fidelity Speech Synthesis. arXiv preprint arXiv:1711.10433, 2017.
[3] SALIMANS, Tim; KINGMA, Diederik P. Weight normalization: A simple reparameterization to accelerate training of deep neural networks. In: Advances in Neural Information Processing Systems. 2016. p. 901-909.

**Comment by beam2d** [comment]
I think the (exponential moving average) Polyak Averaging could be best implemented as an optimizer post hook in Chainer. It should have a clone of the target link and accumulates the parameters after each optimizer update.

**Comment by dhgrs** [comment]
If EMA is implemented as a post hook, EMA parameters are not used during validation by evaluator, aren't they?
My implementation by classifier-like wrapper uses EMA parameters during validation.
Which is better, do you think?

**Comment by beam2d** [comment]
You can pass an arbitrary link object to Evaluator. If you make a clone of the target link and update its parameter using EMA post hook, you can pass that cloned one instead of the original one to Evaluator to do validation with the EMA parameters.

**Comment by iwiwi** [comment]
I think it would be nice to have this feature in Chainer. May I reopen this issue? Please close if inappropriate. Thanks!
